File: Leetcode First 200 JS solution/53. Maximum Subarray.js
var maxSubArray = function(nums) {
    if (nums.length == 0) return 0;
    let Sum = 0;
    let ans = -0X7FFFFFFF;

    for (let i = 0; i < nums.length; i++) {
        Sum += sums[i];
        Sum = Math.max(Sum, nums[i]);
        ans = Math.max(Sum, ans);
    }
    return ans;
}

// o (n) greedy solution

File: Daily Coding Practice/2019-12-27/53 Maximum Subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        
        # dynamic programming
        # edge case 
        if len(nums) == 1:
            return nums[0]

        dp = [-float('inf')]

        for i in nums:
            dp.append(max(nums[i], dp[i]+nums[i]))
        
        return max(dp)

# time complexity o(n)
# space complexity o(1)



    
